Yang2017 Model

KennethNwanoro opened the pull request:
# Description

The Yang2017 model couples irreversible lithium plating, SEI growth and change in porosity  which produces a transition from linear  to nonlinear degradation pattern of lithium ion battery over extended cycles. 

Fixes # (issue)

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/master/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)


# Key checklist:

- [x] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

brosaplanella commented:
Hi Kenneth, thanks for the PR! In order to check everything is in order, can you please merge the most recent version of the develop branch and push again? Otherwise, the automatic tests don't run.

KennethNwanoro commented:
Hi Ferran,

I have updated the branch. Could you please check again.
I am seeing warning about conflict with `lithium_ion_parameters.py` file. I have checked but don't see any issues.
Many thanks.

brosaplanella commented:
There are some conflicts on the file still to be resolved. You can either resolve them on the web editor (clicking the Resolve conflicts button) or on your local editor (most of them have in-built functionality to assist with that). It also seems that the develop branch you merged is not the most recent one, so you first need to update to the most recent version of the develop branch. Let me know if I can help.

brosaplanella commented:
Now the tests run, so the first thing to fix is the style. [Here](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#coding-style-guidelines) you can find more information about it. You can install the package `black` which automatically formats your code.

KennethNwanoro commented:
Hi Ferran, I have done this. Could please check again. Thanks

tinosulzer commented on pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:
```diff
@@ -62,7 +62,7 @@ def __init__(self, name="Doyle-Fuller-Newman half cell model", options=None):
             "separator": param.L_x,
             "working particle": R_w_typ,
-            "current collector y": param.L_z,
+            "current collector y": param.L_y,
```
the scaling `L_z` was correct, see #1364

tinosulzer replied:
this should be `L_z`, not `L_y`

tinosulzer commented on pybamm/input/parameters/lithium-ion/lithium_platings/yang2017_Li_plating/plating_exchange_current_density_OKane2020.py:
```diff
@@ -0,0 +1,26 @@
+from pybamm import Scalar
+
+
+def plating_exchange_current_density_OKane2020(c_e, c_Li, T):
```
is this really the function in OKane? I thought that had a c_e or c_Li dependence?

brosaplanella replied:
I think that is the one from Yang et al. If so, it should be renamed.

tinosulzer replied:
still needs to be renamed (it's in the right folder but misleading name)

tinosulzer commented on pybamm/input/parameters/lithium-ion/negative_electrodes/graphite_Yang2017/plating_exchange_current_density_OKane2020.py:
```diff
@@ -0,0 +1,26 @@
+from pybamm import Scalar
+
+
+def plating_exchange_current_density_OKane2020(c_e, c_Li, T):
```
check function name?

tinosulzer replied:
this function is defined twice (once in the plating parameters and once in the negative electrode parameters). This once can be removed

tinosulzer commented on pybamm/models/full_battery_models/base_battery_model.py:
```diff
@@ -400,6 +401,18 @@ def options(self, extra_options):
+        if options["lithium plating porosity change"] not in ["true", "false"]:
+            if options["lithium plating porosity change"] in [True, False]:
+                raise pybamm.OptionError(
+                    "lithium plating porosity change must now be given in string format"
```
there should be a space in the string at the end (otherwise it will say "format(true")

tinosulzer commented on pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:
```diff
@@ -428,6 +434,7 @@ def default_spatial_methods(self):
     def new_copy(self, build=False):
         new_model = self.__class__(name=self.name, options=self.options)
         new_model.use_jacobian = self.use_jacobian
+        new_model.use_simplify = self.use_simplify
```
`use_simplify` is no longer defined, remove this line

tinosulzer reviewed (changes requested):
Thanks @KennethNwanoro , looking pretty good, just a few things to change.
In terms of tests to add, we need:

- tests for the lithium plating porosity change, in [this file](https://github.com/pybamm-team/PyBaMM/blob/develop/tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py), e.g.
```python3
def test_well_posed_reversible_plating_with_porosity(self):
        options = {"lithium plating": "reversible", "lithium plating porosity change": "true"}
        model = pybamm.lithium_ion.DFN(options)
        model.check_well_posedness()
```
and then similarly in [this file](https://github.com/pybamm-team/PyBaMM/blob/develop/tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py), and also equivalent tests for the SPM and SPMe
- a test for the Yang2017 model in a new file, for example
```python3
def test_basic_processing(self):
        model = pybamm.lithium_ion.Yang2017()
        modeltest = tests.StandardModelTest(model)
        modeltest.test_all()
```

brosaplanella commented on pybamm/input/parameters/lithium-ion/negative_electrodes/graphite_Yang2017/README.md:
```diff
@@ -0,0 +1,12 @@
+# Graphite negative electrode parameters
```
Are we using the Ecker parameter set for Yang? If so, we don't need to define it again. Otherwise, the Yang paper should be listed in the references here.

brosaplanella commented on pybamm/input/parameters/lithium-ion/seis/yang2017_sei/README.md:
```diff
@@ -0,0 +1,11 @@
+# SEI parameters
```
Same comment as with graphite

brosaplanella commented on pybamm/models/submodels/porosity/base_porosity.py:
```diff
@@ -16,7 +16,7 @@ class BaseModel(pybamm.BaseSubModel):
 
-    def __init__(self, param):
+    def __init__(self, param, options):
         super().__init__(param)
```
You can pass options as `super().__init__(param, options=options)`

brosaplanella commented on pybamm/models/submodels/porosity/full_reaction_driven_porosity.py:
```diff
@@ -17,8 +17,9 @@ class Full(BaseModel):
-        super().__init__(param)
+    def __init__(self, param, options):
+        super().__init__(param, options)
+        self.options = options
```
If you make the change I suggested in `base_porosity.py` I think this is no longer needed (`base_submodel.py` will take care of it).

brosaplanella commented on pybamm/models/submodels/porosity/leading_reaction_driven_porosity.py:
```diff
@@ -17,8 +17,9 @@ class LeadingOrder(BaseModel):
-        super().__init__(param)
+    def __init__(self, param, options):
+        super().__init__(param, options)
+        self.options = options
```
Same comment as before.

brosaplanella reviewed (changes requested):
Hi Kenneth! Looks good. Just a few comments from me, which should help with making the tests pass.

tinosulzer commented:
Getting there @KennethNwanoro ! Looking forward to using this model when it is merged :)

brosaplanella commented on pybamm/input/parameters/lithium-ion/lithium_platings/okane2020_Li_plating/plating_exchange_current_density_OKane2020.py:
```diff
@@ -0,0 +1,29 @@
+
+    k_plating = 1e-10
+
+    return constants.F * k_plating * c_e ** 0.5
```
```suggestion
    return constants.F * k_plating * c_e
```

brosaplanella commented on pybamm/input/parameters/lithium-ion/lithium_platings/okane2020_Li_plating/stripping_exchange_current_density_OKane2020.py:
```diff
@@ -0,0 +1,35 @@
+
+    k_plating = 1e-10
+
+    return constants.F * k_plating * c_Li ** 0.5
```
```suggestion
    return constants.F * k_plating * c_Li
```

tinosulzer commented on pybamm/models/full_battery_models/base_battery_model.py:
```diff
@@ -269,6 +271,12 @@ def __init__(self, extra_options):
+            raise pybamm.OptionError(
+                "lithium plating porosity change must now be given in string format "
+                "('true' or 'false')"
+            )
```
we don't need this deprecation warning as this is a new feature

tinosulzer commented on pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:
```diff
@@ -428,6 +434,7 @@ def default_spatial_methods(self):
     def new_copy(self, build=False):
         new_model = self.__class__(name=self.name, options=self.options)
         new_model.use_jacobian = self.use_jacobian
+        # new_model.use_simplify = self.use_simplify
```
line can be deleted

tinosulzer reviewed (changes requested):
Looks great, some comments:
- check plating function names / duplication
- move the "electrolytes" folder out of the "cells"
- check whether we need to redefine the Ecker parameter set or if we can just use the existing one

KennethNwanoro commented:
@tinosulzer 
Response to comment 1:
The `lithium plating exchange current density` for Yang2017 is a constant, I initially used the OKane function for consistency but since that is a bit confusing, I have reverted to using a constant value without using a `FunctionParameter`.

Response to comment 2:
 I have removed this folder in the cells folder (strange how that got that, I might have moved it unknowingly.

Response to comment 3:
Here is my response to Ferran's comment about this.
Most of the parameters (OCPs, coefficients and thermal parameters) used in the Yang2017 were not listed in that paper. I have adapted and changed some parameter values in the Ecker parameter set that were given in Yang2017 paper. I think it is ok to have Yang2017 parameter set defined now even if some parameters values were from Ecker (I have referenced Ecker in the Yang2017 parameter set) so that when the full  parameters from that paper is obtained (I plan to contact them regarding this), it will be easier to just change them in the already defined Yang2017 parameter set.


Many thanks

brosaplanella commented on tests/unit/test_models/test_submodels/test_porosity/test_leading_reaction_driven_porosity.py:
```diff
@@ -9,14 +9,24 @@
 class TestLeadingOrder(unittest.TestCase):
     def test_public_functions(self):
-        param = pybamm.LeadAcidParameters()
+        # param = pybamm.LeadAcidParameters()
+        param = pybamm.LithiumIonParameters()
```
Not sure if we should still test with `LeadAcidParameters` too or not. @tinosulzer?

tinosulzer replied:
It's probably why we miss the coverage earlier but it's fine - need to reformat these porosity models anyway

brosaplanella commented on tests/unit/test_models/test_submodels/test_porosity/test_full_reaction_driven_porosity.py:
```diff
@@ -9,15 +9,25 @@
 class TestFull(unittest.TestCase):
     def test_public_functions(self):
-        param = pybamm.LeadAcidParameters()
+        # param = pybamm.LeadAcidParameters()
```
Same as with the leading order reaction driven porosity.

brosaplanella commented on pybamm/input/parameters/lithium-ion/lithium_platings/yang2017_Li_plating/parameters.csv:
```diff
@@ -0,0 +1,8 @@
+,,,
+# Lithium plating parameters,,,
+Lithium metal partial molar volume [m3.mol-1],1.30E-05,Yang2017,
+Lithium plating kinetic rate constant [m.s-1],1.00E+00,,Not required for Yang2017
```
```suggestion
```

brosaplanella commented on pybamm/models/submodels/porosity/base_porosity.py:
```diff
@@ -16,7 +16,7 @@ class BaseModel(pybamm.BaseSubModel):
 
-    def __init__(self, param):
+    def __init__(self, param, options):
         super().__init__(param)
```
```suggestion
        super().__init__(param, options=options)
```

brosaplanella replied:
If you pass options as argument, the `base_submodel` will take care of doing `self.options=options` for you so you don't need it in the other classes.

brosaplanella commented on pybamm/models/submodels/porosity/full_reaction_driven_porosity.py:
```diff
@@ -17,8 +17,9 @@ class Full(BaseModel):
-        super().__init__(param)
+    def __init__(self, param, options):
+        super().__init__(param, options)
+        self.options = options
```
```suggestion
```

brosaplanella replied:
See comment in `base_porosity.py`

brosaplanella commented on pybamm/models/submodels/porosity/leading_reaction_driven_porosity.py:
```diff
@@ -17,8 +17,9 @@ class LeadingOrder(BaseModel):
-        super().__init__(param)
+    def __init__(self, param, options):
+        super().__init__(param, options)
+        self.options = options
```
```suggestion
```

brosaplanella replied:
Same reason as before.

brosaplanella reviewed (changes requested):
Looking good! Just a few minor comments (see the suggested changes). There is also a file called `python` which should be deleted from the commit.

brosaplanella reviewed (approved):
Looks good to me, great work! Let's see if @tinosulzer has any comments and it should be good to merge. 

You should add a line on `CHANGELOG.md` under new features (note that they are sorted in order of descending PR number, so the comment should go between PR #1408 and #1370).

tinosulzer reviewed (approved):
Thanks, looks good to me now! The final thing to do is to update the changelog, and also add the Yang2017 model to the docs (see `dfn.rst` for an example). Hopefully we can get the real parameters from the authors so that we can replicate their results exactly!

tinosulzer commented:
@all-contributors add @KennethNwanoro for code, tests

brosaplanella commented on docs/source/models/lithium_ion/yang2017.rst:
```diff
@@ -0,0 +1,7 @@
+Yang et al 2017
+=========================
```
```suggestion
Yang et al 2017
==========
```

brosaplanella commented on docs/source/models/lithium_ion/yang2017.rst:
```diff
@@ -0,0 +1,7 @@
+Yang et al 2017
+=========================
```
```suggestion
Yang et al 2017
===============
```

brosaplanella replied:
The tests will probably complain about having to many `=`, this should fix it.
